src/rdata.cpp

```diff
@@ -404,8 +406,46 @@ void ReturnData::processBackwardProblem(ForwardProblem const &fwd,
     auto xB = bwd.getAdjointState();
     auto xQB = bwd.getAdjointQuadrature();
 
-    /* NB: This nested loop will not be necessary for fully adjoint
-       preequilibration or post-equilibration without further time points */
+    if (preeq && preeq->hasQuadrature()) {
+        handleSx0Backward(model, *preeq, xQB);
+    } else {
+        handleSx0Forward(model, llhS0, xB);
+    }
+
+    for (int iJ = 0; iJ < model.nJ; iJ++) {
+        for (int ip = 0; ip < model.nplist(); ip++) {
+            if (iJ == 0) {
+                sllh.at(ip) -= llhS0[ip] + xQB[ip * model.nJ];
+            } else {
+                s2llh.at(iJ - 1 + ip * (model.nJ - 1)) -=
+                    llhS0[ip + iJ * model.nplist()] + xQB[iJ + ip * model.nJ];
+            }
+        }
+    }
+}
+
+void ReturnData::handleSx0Backward(const Model &model,
+                                   SteadystateProblem const &preeq,
+                                   AmiVector &xQB) const {
+    /* If preequilibration is run in adjoint mode, the scalar product of sx0
+       with its adjoint counterpart (see handleSx0Forward()) is not necessary:
+       the actual simulation is "extended" by the preequilibration time.
+       At initialization (at t=-inf), the adjoint state is in steady state (= 0)
```

Comment:
this only works if the the jacobian has full rank right? Otherwise the adjoint state could be any vector from the null space?

Reply:
In principle, this will work always, not depending on the Jacobian.
What depends however on the Jacobian having full rank is the computation of `xQBpreeq`. The analytical solution is indeed *only* available for full-rank systems. But a solution based on numerical integration will always work (as long as passing xB from BackwardProblem to SteadyStateProblem is possible, which is currently not the case if solver states must be reinitialized).